src/components/DropdownMenu/DropdownMenu.tsx

```diff
@@ -1,211 +1,171 @@
-                    size={size}
-                    menuProps={menuProps}
-                >
-                    {children}
-                </DropdownMenuPopup>
-            </React.Fragment>
-        );
-    }
-
-    private renderDefaultSwitcher() {
-        const {size, defaultSwitcherClassName, defaultSwitcherProps, icon} = this.props;
-
-        return (
-            <Button
-                view="flat"
-                size={size}
-                {...defaultSwitcherProps}
-                className={b('switcher-button', defaultSwitcherClassName)}
-                disabled={this.props.disabled}
-            >
-                {icon}
-            </Button>
-        );
-    }
-
-    private onMenuItemClick = (
-        event: React.MouseEvent<HTMLElement, MouseEvent>,
-        action: DropdownMenuItemAction<T>,
-    ) => {
-        action(event, this.props.data);
-        this.props.onMenuToggle();
-        this.hidePopup();
-    };
-
-    private onSwitcherClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
-        if (this.props.disabled) {
+    /**
+     * Custom content inside the menu popup.
+     */
+    children?: ReactNode;
+};
+
+export const DropdownMenu = <T,>({
+    items = [],
+    size = 'm',
+    icon = <Icon data={DotsIcon} />,
+    onMenuToggle,
+    hideOnScroll = true,
+    data,
+    disabled,
+    switcher,
+    switcherWrapperClassName,
+    defaultSwitcherProps,
+    defaultSwitcherClassName,
+    onSwitcherClick,
+    menuProps,
+    popupClassName,
+    popupPlacement,
+    children,
+}: DropdownMenuProps<T>) => {
+    const [isPopupShown, setPopupShown] = useState(false);
+    const anchorRef = useRef<HTMLDivElement | null>(null);
+
+    const handleSwitcherClick: MouseEventHandler<HTMLDivElement> = (event) => {
+        if (disabled) {
             return;
         }
 
-        this.props.onMenuToggle();
-
-        if (this.props.onSwitcherClick) {
-            this.props.onSwitcherClick(e);
-        }
-
-        if (this.state.popupIsShowed) {
-            this.hidePopup();
-        } else {
-            this.showPopup();
-        }
-    };
-
-    private onScroll = (event: Event) => {
-        let element: null | (Node & ParentNode) = this.anchorRef.current;
-
-        while (element && element !== event.target) {
-            element = element.parentNode;
-        }
-
-        if (element === event.target) {
-            this.props.onMenuToggle();
-            this.hidePopup();
-        }
+        onMenuToggle?.();
+        onSwitcherClick?.(event);
+        setPopupShown((value) => !value);
     };
 
-    private showPopup() {
-        this.setState({popupIsShowed: true});
-
-        if (this.props.hideOnScroll) {
-            document.addEventListener('scroll', this.onScroll, true);
+    const handleMenuItemClick = useCallback(
+        (event: React.MouseEvent<HTMLElement, MouseEvent>, action: DropdownMenuItemAction<T>) => {
+            action(event, data);
+            onMenuToggle?.();
```

Comment:
onMenuToggle is called when display state changes, i think it is ok to move it to separate `useEffect`

```js
useEffect(() => {
  onMenuToggle?.();
}, [isPopupShown]);
```

Reply:
such an effect will also run right after the component has mounted, we'll have to handle it to prevent an unnecessary `onMenuToggle()` call; I think it's ok to leave it as it is